### ISSUE TITLE
New protocol icons

### ISSUE THREAD

**Issue by kaitmore** [DecipherNow/gm-ui-components#580]
We need several new icons to represent protocols in support of https://github.com/DecipherNow/gm-dashboard/issues/3642

![image](https://user-images.githubusercontent.com/22768918/91578539-50110100-e918-11ea-851c-178def6ce176.png)

Full mocks: https://app.lucidchart.com/documents/edit/2870201a-3378-4519-b42f-79d5d355da8c/0_0

**Comment by tilleryd** [comment]
SPIFFE is a standard that provides cryptographic identities that can be used when establishing TLS connections. I like that we are thinking about SPIFFE but if it is being used in the mesh I think we should just ensure that TLS is indicated as the protocol.

**Comment by chrismholmes** [comment]
TLS is not a separate protocol.  It should be indicated if its enabled or not -- but we should still show that protocol in use for traffic flow (HTTP 1.1/HTTP 2.0, etc. ) where are the rest of the protocols?  Are we going to slap TCP on everything else even though we should know if its specifically MongoDB, Kafka, REDIS, PostgresDB?

**Comment by joshua-rutherford** [comment]
To close out this conversation:

1. The engineering team will ensure that the frontend can display zero or more protocols per card.  
1. The protocols will not be hardcoded in the frontend, rather will be based upon a string array provided by the back end.
1. The backend will map from mesh configurations to frontend rendered text (e.g., type.googleapis.com/envoy.extensions.filters.network.mysql_proxy.v3.MySQLProxy -> MySQL).
1. The backend will default to "UNKNOWN" should a mapping not exist for any protocols in use.
1. Extending the mappings of mesh configurations to frontend rendered text should not require recompilation of the backend (i.e., they must be configurable at deployment or run time).

@chrismholmes, does this sound like the requirements you had in mind?

**Comment by chrismholmes** [comment]
The above are mostly technical details.

The functional/product owner requirement:

* Dashboard Updates
    * Visualize the type of service protocol such as:
        * (HTTP vs. TCP vs. Kafka vs. MongoDB vs. x)

Approved in the release cycle for 1.3 are these mocks: https://app.lucidchart.com/documents/edit/2870201a-3378-4519-b42f-79d5d355da8c/0_0#

I believe these mocks handle the intent of the requirement very well. You would see visually on the card the protocol or protocols.  The crux of the "functional" requirement is #3 above in your list. I am unclear which network filters/proxies we actually support, but we need to label the ones we know and do which I think are documented here:

https://docs.greymatter.io/usage/fabric/filters/network#envoy-network-filters

Of note on this list as far as "protocol", I think they conflate the protocol chatter with metadata about the chatter in their examples and docs.  As example:

- Mongo
- Redis
- HTTP
- GRPC
- WS
- Kafka
- RDMBS (MySql)
- Dubbo
- Echo
- Zookeeper
- Thrift
- TCP

Are all "protocol" labels (they are all TCP at the end of the day).  They are refined mechanisms and added overlay information specific to the thing you are talking with on the backend.

Items such as:

- Local rate limit
- Rate limit
- SNI
- RBAC (in the case of network filters see our docs)
- Network External Auth
- TLS
- Direct Réponse

Seem to be metadata about the connection and more than what/how you talk to it.  So I would assume they should be treated differently.

Would it be better to fall back on TCP vs. UNKOWN?  Since everything is ultimately TCP until we truly support UDP - which we do not currently irregardless if Envoy does or not.

You lost me on #5 in the list.  That seems like a technical detail and if you believe that is the prudent how, I will defer to you on that.

**Comment by joshua-rutherford** [comment]
My point on number five was to remove the requirement for a new catalog release should a new network filter be introduced.  Instead this becomes a deployment concern. If you are fine with the limitation that new labels will only be added via releases then this can be ignored.

Regarding `UNKNOWN` protocols, I think we should probably do the following:

1. If at least one known protocol filter is present, we display the known protocols and ignore the unknown.
2. If no known protocol filters are present we display `UNKNOWN`.

I concur that displaying the non-protocol filters as `UNKNOWN` or even as `RBAC` is not useful.  But I do not think assuming `TCP` is the right answer either. In the situation where the listener is misconfigured (e.g., doesn't have any filters listed) we would be indicated an invalid state.  We should avoid assumptions that can lead to misreported states at all costs.

**Comment by chrismholmes** [comment]
This is where I am just not really how things work and our documentation doesn't really display usage.  Meaning...

To setup network filters in JSON there is a block like so:

"network_filters": {
{
  "zone_key": "default-zone",
  "listener_key": "listener2",
  "name": "listener2",
  "domain_keys": [
    "domain1"
  ],
  "ip": "0.0.0.0",
  "port": 8080,
  "protocol": "http_auto",
  "active_network_filters": [
    "envoy.tcp_proxy"
  ],
  "network_filters": {
    "envoy_tcp_proxy": {
      "stat_prefix": "tcp_proxy",
      "cluster": "tcp_proxy"
    }
  }
}

I would imagine you can't have just:

- Local rate limit
- Rate limit
- SNI
- RBAC (in the case of network filters see our docs)
- Network External Auth
- TLS
- Direct Réponse

without specifying a service or a protocol you are using it for - but I don't understand how these configs get wired together (which is a disconnect I have when really explaining this stuff).

and that you would have to have something more like below where the filters are chained together? (I'm sure my configs are messed up...but neither our docs or Envoy's really walk you through real world configuration).

{
  "zone_key": "default-zone",
  "listener_key": "listener2",
  "name": "listener2",
  "domain_keys": [
    "domain1"
  ],
  "ip": "0.0.0.0",
  "port": 8080,
  "protocol": "http_auto",
  "active_network_filters": [
    "envoy.tcp_proxy"
  ],
  "network_filters": {
    "envoy_tcp_proxy": {
      "stat_prefix": "tcp_proxy",
      "cluster": "tcp_proxy"
    }
  "envoy_rate_limit": {
      "stat_prefix": "edge",
      "domain": "edge",
      "failure_mode_deny": true,
      "descriptors": [
        {
          "entries": [
            {
              "key": "path",
              "value": "/"
            }
          ]
        }
      ],
      "rate_limit_service": {
        "grpc_service": {
          "envoy_grpc": {
            "cluster_name": "ratelimit"
          }
        }
      }
  }
}

**Comment by joshua-rutherford** [comment]
The current implementation for network filters in Control API allows anything to be specified as long as it's valid JSON: 

https://github.com/DecipherNow/gm-control-api/blob/release-1.5/api/listener.go#L133

Control does "validate" the JSON when it attempts to deserialize it but there is no indication in the Control API as to whether this is successful.  Further, the real validity of the filter chain would not be know until Envoy acks or nacks the configuration which is yet one more step removed from the Control API.  

In short, our configurations are a bit of an open loop right now with no positive feedback.  There are advantages and disadvantages to this approach but one of the disadvantages is that we perform no real validation until the sidecar.

All this to say that a user does not know that the provided configuration is invalid at the time the push it so we can and will end up with situations as I described.
